Source/WebCore/loader/ResourceMonitorChecker.cpp

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "ResourceMonitorChecker.h"
+
+#include <wtf/CryptographicallyRandomNumber.h>
+#include <wtf/StdLibExtras.h>
+
+namespace WebCore {
+
+#if ENABLE(CONTENT_EXTENSIONS)
+
+ResourceMonitorChecker& ResourceMonitorChecker::singleton()
+{
+    static NeverDestroyed<ResourceMonitorChecker> globalChecker;
+    return globalChecker;
+}
+
+ResourceMonitorChecker::ResourceMonitorChecker()
+    : m_workQueue { WorkQueue::create("ResourceMonitorChecker Work Queue"_s) }
+{
+}
+
+void ResourceMonitorChecker::checkEligibility(const ContentExtensions::ResourceLoadInfo& info, CompletionHandler<void(Eligibility)>&& completionHandler)
+{
+    ASSERT(isMainThread());
+
+    ContentExtensions::ResourceLoadInfo copy = {
+        .resourceURL = info.resourceURL.isolatedCopy(),
+        .mainDocumentURL = info.mainDocumentURL.isolatedCopy(),
+        .frameURL = info.frameURL.isolatedCopy(),
+        .type = info.type,
+        .mainFrameContext = info.mainFrameContext,
+    };
+
+    protectedWorkQueue()->dispatch([this, info = WTFMove(info), completionHandler = WTFMove(completionHandler)] mutable {
```

Comment:
I think the usual pattern for something like this would be to define `isolatedCopy() const` and `isolatedCopy &&` on `ContentExtensions::ResourceLoadInfo`. Then when you want to pass that object across threads, do `info = crossThreadCopy(info)` or `info = crossThreadCopy(WTFMove(info))`.

Reply:
Correct. info needs an isolatedCopy() function and we should call `info = crossThreadCopy(WTFMove(info))` here.